### PR TITLE
notifications

### PR DESCRIPTION
**Pull request by yenda**

### CODE REVIEW

File: docs/draft/notifications.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+An early implementation of push notifications was done using the Firebase library which required users to exchange notifications ids and send notifications through Firebase. This was fairly limited by the fact that one could only send/receive a notification after exchanging these ids with the contact. We were also not satisfied with the fact that it was a huge dependency, which required the installation of `Play Services` that some users weren't comfortable with. 
+
+## Local Notifications
+
+Using local notifications, we can offer notifications for any event in the app without any privacy cost and dependency on third party services.
+
+### Android
+
+Android allows running services on the device. When the user enables notifications, the client starts a ``Foreground Service`, and a permanent notification indicating that the service is running is displayed to the user, as required by Android guidelines.
```
Comment (decanus) [inline docs/draft/notifications.md]:
```suggestion
Android allows running services on the device. When the user enables notifications, the client starts a `Foreground Service`, and a permanent notification indicating that the service is running is displayed to the user, as required by Android guidelines.
```

**Review by decanus — CHANGES_REQUESTED** [review]

File: docs/draft/notifications.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
+---
+layout: default
+title: Notifications
+nav_order: 2
+has_children: true
+permalink: /specs/draft
```
Comment (oskarth) [inline docs/draft/notifications.md]: Add number to spec and PR title; make sure it doesn't conflict with existing PRs

File: docs/draft/notifications.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+---
+
+# Notifications in Status
+
+## Why is there no Push Notifications?
```
Comment (oskarth) [inline docs/draft/notifications.md]: It'd be better to start by describing how notifications work (to the extent that they do). Then we can have this as an FAQ/motivation for now doing X.

I'd also cut some of the "earlier implementations..." as it isn't useful in a spec. If we want to keep this, it should be in a later section, like FAQ etc.

File: docs/draft/notifications.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+Push Notifications, as offered by Apple and Google are a privacy concern, they require a centralized service that is aware of who the notification needs to be delivered to.
+
+An early implementation of push notifications was done using the Firebase library which required users to exchange notifications ids and send notifications through Firebase. This was fairly limited by the fact that one could only send/receive a notification after exchanging these ids with the contact. We were also not satisfied with the fact that it was a huge dependency, which required the installation of `Play Services` that some users weren't comfortable with. 
+
+## Local Notifications
```
Comment (oskarth) [inline docs/draft/notifications.md]: This is a better start. I'd start by describing requirements and design goals. Then lead to how we solve it with local notifications for Android, and write about limitation for iOS.

File: docs/draft/notifications.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+Using local notifications, we can offer notifications for any event in the app without any privacy cost and dependency on third party services.
+
+### Android
+
+Android allows running services on the device. When the user enables notifications, the client starts a `Foreground Service`, and a permanent notification indicating that the service is running is displayed to the user, as required by Android guidelines.
```
Comment (oskarth) [inline docs/draft/notifications.md]: This currently reads more like "the (one) client does this". Ideally it should read something like: "If a client wants to support notifications, it MUST do X, Y, Z etc.

Then less emphasis on specific implementation details, other than what is relevant for all implementations (e.g. Nim based, greenfield, other Go impl, desktop one, etc etc)

File: docs/draft/notifications.md [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+The signals are passed to `NewMessageSignalHandler` which takes care of managing the local notifications. It keeps track of the current active notifications so that new messages are displayed together by chat as it is done in most common chat applications.
+
+Notifications are a separate process in Android, and interaction with a notification generates an `Intent`. To handle intents, the `NewMessageSignalHandler` uses a `BroadcastReceiver`, in order to update the state of local notifications when the user dismisses or tap a notification. If the user taps on a notification, the `BroadcastReceiver` generates a new intent to open the app and uses universal links to get the user to the right place.
+
+
+### iOS
+
+We are not able to offer local notifications on iOS because there is no concept of services in iOS. It offers background updates but they’re not consistently triggered, and cannot be relied upon. The system decides when the background updates are triggered and the heuristics aren't known.
+
+## Future development
```
Comment (oskarth) [inline docs/draft/notifications.md]: I'd de-emphasize this as it isn't the core part of a spec. If we want to include it I'd instead refer to an issue where future development is tracked.

**Review by oskarth — CHANGES_REQUESTED** [review]
Good start! I left some comments in-line. The main thing is shifting mindset from thinking "current app" to "how would a new client implement Status notifications? This has implications for language (MUST, SHOULD, etc) what we include and don't (future development, historical accidents)

File: docs/draft/notifications.md [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+We can and should implement more types of local notifications on Android so that users can be notified for instance when:
+- they are mentioned in any chat, 
+- a keyword they are following appears in any chat, 
+- they receive a transaction or a transaction they sent has been confirmed,
+- the total value of their assets in the wallet reached a threshold
+
+### Notifications for iOS
+
+Since iOS cannot reliably display local notifications, we need to find an alternative. But we should not impact Android users with iOS limitations.
```
Comment (oskarth) [inline docs/draft/notifications.md]: ## Copyright

Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).

File: docs/draft/notifications.md [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+## Local Notifications
+
+Using local notifications, we can offer notifications for any event in the app without any privacy cost and dependency on third party services.
+
+### Android
+
+Android allows running services on the device. When the user enables notifications, the client starts a `Foreground Service`, and a permanent notification indicating that the service is running is displayed to the user, as required by Android guidelines.
+The service simply keeps the app from being killed by the system when it is in the background.
+
+In the Android client, events from `status-go` are handled by the `handleSignal` method of `im.status.ethereum.module`. Those signals are passed to the clojurescript part of the client. This is the general behavior when the client is in the foreground. For the local notifications, we also react on these signals when the app is in the background. Currently, we only display a notification for 1-1 messages, but support for any signal that status-go may send can be easily added.
```
Comment (decanus) [inline docs/draft/notifications.md]:
```suggestion
In the Android client, events from `status-go` are handled by the `handleSignal` method of `im.status.ethereum.module`. Those signals are passed to the clojurescript part of the client. This is the general behavior when the client is in the foreground. For the local notifications, we also react to these signals when the app is in the background. Currently, we only display a notification for 1-1 messages, but support for any signal that status-go may send can be easily added.
```

**Comment by oskarth** [comment]
Ping @yenda

**Review by decanus — CHANGES_REQUESTED** [review]

**Comment by oskarth** [comment]
Ping @yenda whats the state of this? It'd be great if we can merge it and create follow up issues where things still need to be done.

**Comment by oskarth** [comment]
Seems like some of the earlier outstanding comments are still unresolved? @yenda

**Review by oskarth — APPROVED** [review]

**Comment by yenda** [comment]
@andremedeiros @decanus is this all good for you?

**Review by andremedeiros — APPROVED** [review]

**Review by decanus — APPROVED** [review]

**Pull request merged**